src/inmanta/deploy/scheduler.py

```diff
@@ -1064,29 +1072,34 @@ async def deploy_start(self, action_id: uuid.UUID, resource: ResourceIdStr) -> O
+        state: Optional[ResourceState] = None
         try:
-            state = await self._update_scheduler_state_for_finished_deploy(intent.details.attribute_hash, result, finished)
-        except StaleResource:
-            # The resource is no longer managed (or in rare cases it has shortly become unmanaged sometime between this version
-            # and the currently managed version, either way, the deploy that finished represents a stale intent).
-            # We do still want to report that the deploy finished for this version, even if it is stale, we simply don't report
-            # any state.
-            # new, ununtouched state for this resource id), because that still represents the most current state for this
-            # resource, even if it is independent from this deploy.
-            # The scheduler will then report that this resource id with this specific version finished deploy.
-            state = None
-        # Write deployment result to the database.
-        await self.state_update_manager.send_deploy_done(
-            attribute_hash=intent.details.attribute_hash,
-            result=result,
-            state=state,
-            started=intent.deploy_start,
-            finished=finished,
-        )
-        async with self._scheduler_lock:
-            # report to the scheduled work that we're done. Never report stale deploys.
-            self._work.finished_deploy(result.resource_id)
+            try:
+                state = await self._update_scheduler_state_for_finished_deploy(intent.details.attribute_hash, result, finished)
+            except StaleResource:
+                # The resource is no longer managed (or in rare cases it has shortly become unmanaged sometime
+                # between this version and the currently managed version, either way, the deploy that finished
+                # represents a stale intent). We do still want to report that the deploy finished for this version,
+                # even if it is stale, we simply don't report any state. new, ununtouched state for this resource
+                # id), because that still represents the most current state for this resource, even if it is
+                # independent from this deploy. The scheduler will then report that this resource id with this
+                # specific version finished deploy.
+                state = None
+            # Write deployment result to the database.
+            await self.state_update_manager.send_deploy_done(
+                attribute_hash=intent.details.attribute_hash,
+                result=result,
+                state=state,
+                started=intent.deploy_start,
+                finished=finished,
+            )
+        finally:
+            # Always do this, even if the DB is broken
+            async with self._scheduler_lock:
+                if state is not None:
+                    self._timer_manager.update_timer(intent.details.resource_id, state=state)
```

Comment:
I think we probably want to fetch the latest state here, to make sure the timer is accurate.

In which case, the `state` declaration can move into the first try.